### PR TITLE
Dasherise member names

### LINKED ISSUE #13: JSON API dasherize property names
_From @thhermansen on November 30, 2015 14:31_

Hi,

I wasn't able to figure this one out myself. Given the following:

``` ruby
class User < Roar::Decorator
  include Roar::JSON::JSONAPI

  property :full_name
end
```

Is there a way to make the serialised version of a user to have property named `full-name`?

I can solve it by using `#property`'s `as`-option, but I'd like to dasherize all property names automatically.

_Copied from original issue: trailblazer/roar#173_

### PR DESCRIPTION
**Pull request by myabc**
Dasherise property names.

Closes #13

### CODE REVIEW

File: lib/roar/json/json_api/nesting.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      module Nesting
+        def nested(name, options = {}, &block)
+          options = options.merge(
+            getter:   ->(_opts) { self },
+            setter:   ->(opts)  {},
+            instance: ->(_opts) { self }
+          )
+
+          if block
+            options[:_nested_builder] = nested_builder
```
Comment (myabc) [inline lib/roar/json/json_api/nesting.rb]: Reimplementing the whole method in order to override `options[:_nested_builder]`:
https://github.com/trailblazer/representable/blob/master/lib/representable/declarative.rb#L30

File: lib/roar/json/json_api/nesting.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            options[:_base]           = Decorator.default_nested_class
+          end
+
+          property(name, options, &block)
+        end
+
+        NestedBuilder = ->(options) do
+          Class.new(options[:_base]) do
+            feature(*options[:_features])
+            defaults(&DEFAULTS_AS)
```
Comment (myabc) [inline lib/roar/json/json_api/nesting.rb]: @apotonick Per your comment here https://github.com/trailblazer/roar-jsonapi/issues/13#issuecomment-266230178

> Defaults are not "inherited" into nested representers. I need to think about if that's a good idea or if we should introduce a inline: true option or something.

What do you think about this implementation?

File: lib/roar/json/json_api/member_name.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    module JSONAPI
+      class MemberName
+        LENIENT_FILTER_REGEXP = /([^[:alnum:][-_ ]]+)/
+        STRICT_FILTER_REGEXP  = /([^[0-9a-z][-_]]+)/
+
+        def self.call(name, options = {})
+          new.(name, options)
+        end
+
+        def call(name, options = {})
```
Comment (apotonick) [inline lib/roar/json/json_api/member_name.rb:14]: I like the API! :wink:

File: lib/roar/json/json_api/nesting.rb (hunk 1 above)
Comment (apotonick) [reply lib/roar/json/json_api/nesting.rb]: Why do we need to redefine the `NestedBuilder` ?

Comment (myabc) [reply lib/roar/json/json_api/nesting.rb]: @apotonick because I want `NestedBuilder` that inherits/or sets defaults.

**Pull request merged**